### PR TITLE
Extend httpclientbuilder to allow setting proxy

### PR DESCRIPTION
**Pull request by harishch1998**
Code changes to extend the HTTPClientBuilder to allow setting a proxy on the internal httpclient.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/rest/HTTPClient.java [hunk 1]
```diff
@@ -468,6 +476,12 @@ public Builder uri(String path) {
       return this;
     }
 
+    public Builder withProxy(String hostIp) {
+      Preconditions.checkNotNull(hostIp, "Invalid hostIp for http client proxy: null");
+      this.proxy = new HttpHost(hostIp);
```
Comment (sfc-gh-mparmar) [inline core/src/main/java/org/apache/iceberg/rest/HTTPClient.java:481]: Let's allow setting the port as well

File: core/src/main/java/org/apache/iceberg/rest/HTTPClient.java [hunk 2]
```diff
@@ -468,6 +476,12 @@ public Builder uri(String path) {
       return this;
     }
 
+    public Builder withProxy(String hostIp) {
```
Comment (sfc-gh-mparmar) [inline core/src/main/java/org/apache/iceberg/rest/HTTPClient.java:479]: Could this be a URL instead?

**Review by sfc-gh-mparmar — APPROVED** [review]

**Review by sfc-gh-dhuo — APPROVED** [review]

File: core/src/main/java/org/apache/iceberg/rest/HTTPClient.java [hunk 3]
```diff
@@ -85,6 +86,7 @@ public class HTTPClient implements RESTClient {
 
   private HTTPClient(
       String uri,
+      HttpHost proxy,
```
Comment (sfc-gh-jboeuf) [inline core/src/main/java/org/apache/iceberg/rest/HTTPClient.java:89]: Thanks a lot of the PR!

Can we also add at least support for username/password authentication? Here's how to set it: https://stackoverflow.com/questions/6962047/apache-httpclient-4-1-proxy-authentication

Comment (harishch1998) [reply core/src/main/java/org/apache/iceberg/rest/HTTPClient.java:93]: Done

**Review by sfc-gh-jboeuf — APPROVED** [review]
LGTM, thanks for adding the proxy auth.

**Review by sfc-gh-mparmar — APPROVED** [review]

**Pull request merged**